lib/services/search.js

```diff
@@ -339,17 +313,59 @@ function batch(ops) {
 }
 
 /**
- * Create the ES Client or an empty object
+ * [createClient description]
+ * @param  {[type]} overrideClient [description]
+ * @return {[type]}                [description]
+ */
+function createClient(overrideClient) {
+  return new bluebird(function (resolve, reject) {
+    if (!module.exports.endpoint && !overrideClient) {
+      reject(new Error('No Elastic endpoint or client override'));
+    }
+
+    // Set the exported client
+    module.exports.client = client = module.exports.endpoint && !overrideClient ? new elastic.Client(_.clone(serverConfig)) : _.assign(module.exports.client, overrideClient);
+
+    resolve(client);
+  });
+}
+
+/**
+ * Setup the client and check the health of the client
  *
  * @param {Object} overrideClient
+ * @return {Promise}
  */
 function setup(overrideClient) {
-  if (!module.exports.endpoint && !overrideClient) {
-    throw new Error('No Elastic endpoint or client override');
-  }
+  // Create client + check the connection
+  return createClient(overrideClient)
+    .then(healthCheck)
+    // .then(createInternalIndices)
+    .then(setupMappings)
+    .catch(function (err) {
+      return bluebird.reject(err);
+    });
+}
 
-  // Set the exported client
-  module.exports.client = client = module.exports.endpoint && !overrideClient ? new elastic.Client(_.clone(serverConfig)) : _.assign(module.exports.client, overrideClient);
+/**
+ * Check if the correct indices exist, and if they don't, create them.
+ *
+ * @param {Object} mappings
+ * @returns {Promise}
+ */
+function validateIndices(mappings) {
+  return bluebird.all(_.map(Object.keys(mappings), function (index) {
```

Comment:
The lodash `_.map` is still faster than native map because it uses a for loop, plus it maps to the patterns that we've already established. I'd prefer to leave it until there's a good performance reason to switch them all over.